obs-outputs: Add Mixer FTL support

QuinnDamerell-MS opened the pull request:
This work adds support for Mixer’s custom FTL protocol to OBS. This includes adding support for encoding OPUS audio, adding the FTL submodule, and adding a ftl-output.

Xaymar commented on UI/window-basic-main-outputs.cpp:
```diff
@@ -556,7 +556,7 @@ void SimpleOutput::UpdateStreamingSettings_amd(obs_data_t *settings,
 	obs_data_set_int(settings, "Profile", 100); // High
 	obs_data_set_string(settings, "profile", "high"); // High
-	
+
```
Formatting: Indentation removed for no reason.

notr1ch replied:
This is likely unintentional but probably desired due to the .editorconfig in use for OBS. Not all files are conforming to the .editorconfig format yet, but when saving a modified file, it will enforce some of the rules (in this case, no trailing whitespace) throughout the entire file.

QuinnDamerell-MS replied:
I fixed all of the formating issues. :)

Xaymar commented on UI/window-basic-main-outputs.cpp:
```diff
@@ -565,7 +565,7 @@ void SimpleOutput::UpdateStreamingSettings_amd(obs_data_t *settings,
 	obs_data_set_int(settings, "VBVBuffer", 1);
 	obs_data_set_int(settings, "VBVBuffer.Size", bitrate);
-	
+
```
Formatting: Indentation removed for no reason.

jp9000 replied:
That's trailing whitespace.  His editor probably removed it.  It doesn't really matter if trailing whitespace is removed, probably better that way.

derrod commented on plugins/obs-outputs/ftl-stream.c:
```diff
@@ -0,0 +1,1188 @@
+
+	uint64_t         total_bytes_sent;
+    uint64_t         dropped_frames;
+    uint64_t         last_nack_count;
```
Mixed tabs/spaces

QuinnDamerell-MS replied:
All fixed!

derrod commented on plugins/obs-outputs/ftl-stream.c:
```diff
@@ -0,0 +1,1188 @@
+    }
+    if (os_event_init(&stream->stop_event, OS_EVENT_TYPE_MANUAL) != 0) {
+        goto fail;
+    }
```
Again spaces instead of tabs

QuinnDamerell-MS replied:
All fixed!

derrod commented on plugins/obs-outputs/ftl-stream.c:
```diff
@@ -0,0 +1,1188 @@
+    if (ts) {
+        stream->shutdown_timeout_ts = ts +
+            (uint64_t)stream->max_shutdown_time_sec * 1000000000ULL;
+    }
```
And again

QuinnDamerell-MS replied:
All fixed!

derrod commented on plugins/obs-outputs/ftl-stream.c:
```diff
@@ -0,0 +1,1188 @@
+
+    if (stopping(stream) && ts != 0) {
+        return;
+    }
```
And again

QuinnDamerell-MS replied:
All fixed!

derrod commented on plugins/obs-outputs/ftl-stream.c:
```diff
@@ -0,0 +1,1188 @@
+			speedtest_duration,
+			&results);
+
+    float percent_lost = 0;
```
again...

QuinnDamerell-MS replied:
All fixed!

derrod commented on plugins/obs-outputs/ftl-stream.c:
```diff
@@ -0,0 +1,1188 @@
+    // connection really can't handle the bitrate the user will see either lost frame
+    // and recovered frame counts go up, which is reflect in the dropped_frames count.
+    stream->peak_kbps = stream->params.peak_kbps = user_desired_bitrate * 1.2;
+    ftl_ingest_update_params(&stream->ftl_handle, &stream->params);
```
Again. (This whole section)

QuinnDamerell-MS replied:
All fixed!

derrod commented on plugins/obs-outputs/ftl-stream.c:
```diff
@@ -0,0 +1,1188 @@
+	info("Connection to %s successful", stream->path.array);
+
+	// Always get the peak bitrate when we are starting.
+    set_peak_bitrate(stream);
```
again

QuinnDamerell-MS replied:
All fixed!

derrod commented on plugins/obs-outputs/ftl-stream.c:
```diff
@@ -0,0 +1,1188 @@
+    }
+    if (!obs_output_initialize_encoders(stream->output, 0)) {
+        return false;
+    }
```
again

QuinnDamerell-MS replied:
All fixed!

QuinnDamerell-MS replied:
Fixed.

derrod commented on plugins/obs-outputs/ftl-stream.c:
```diff
@@ -0,0 +1,1188 @@
+
+            // Report nack requests as dropped frames
+            stream->dropped_frames += p->nack_reqs - stream->last_nack_count;
+            stream->last_nack_count = p->nack_reqs;
```
Again

QuinnDamerell-MS replied:
All fixed!

derrod commented on plugins/obs-outputs/ftl-stream.c:
```diff
@@ -0,0 +1,1188 @@
+
+	blog(LOG_WARNING, "ftl-stream: connect thread");
+
+    ret = init_connect(stream);
```
again

QuinnDamerell-MS replied:
All fixed!

derrod commented on plugins/obs-outputs/ftl-stream.c:
```diff
@@ -0,0 +1,1188 @@
+    if (!service)
+    {
+        return OBS_OUTPUT_ERROR;
+    }
```
again

QuinnDamerell-MS replied:
All fixed!

derrod commented on plugins/obs-outputs/ftl-stream.c:
```diff
@@ -0,0 +1,1188 @@
+            blog(LOG_ERROR, "Failed to create ingest handle (%s)",
+                ftl_status_code_to_string(status_code));
+            return OBS_OUTPUT_ERROR;
+        }
```
Again

QuinnDamerell-MS replied:
All fixed!

RytoEX commented on plugins/obs-outputs/ftl-stream.c:
```diff
@@ -0,0 +1,1188 @@
+	if (!pframes) {
+		stream->congestion = (float)buffer_duration_usec /
+			(float)drop_threshold;
+	}
```
Style:  Curly braces are not needed for this `if` statement, since the contents of each branch is a single statement.

QuinnDamerell-MS replied:
Fixed.

RytoEX commented on plugins/obs-outputs/ftl-stream.c:
```diff
@@ -0,0 +1,1188 @@
+
+static void ftl_stream_data(void *data, struct encoder_packet *packet)
+{
+	struct ftl_stream    *stream = data;
```
This struct seems to be meant to be aligned with the declarations below, but it is off by one space when compared to other alignments in this file.

QuinnDamerell-MS replied:
Fixed.

RytoEX commented on plugins/obs-outputs/ftl-stream.c:
```diff
@@ -0,0 +1,1188 @@
+static uint64_t ftl_stream_total_bytes_sent(void *data)
+{
+	struct ftl_stream *stream = data;
+
```
There is a blank newline here that isn't mirrored in similar functions below.  For consistency, I'd suggest removing this newline, or adding one to each of the below functions.

QuinnDamerell-MS replied:
Fixed.

derrod commented on plugins/rtmp-services/data/services.json:
```diff
@@ -1,5 +1,5 @@
 {
-    "format_version": 1,
+    "format_version": 2,
```
This is most likely due to the addition of the ftl type of service not being backwards compatible and thus making this a new version of the format. It also means that older versions of obs won't load the new services file as it could cause issues.

jp9000 replied:
Well they renamed the service entry so it's probably not necessary to increment the format version anymore.

QuinnDamerell-MS commented on UI/window-basic-main-outputs.cpp:
```diff
@@ -689,6 +689,9 @@ bool SimpleOutput::StartStreaming(obs_service_t *service)
+        if (!codec) {
+            return false;
+        }
 
```
All fixed!

QuinnDamerell-MS commented on UI/window-basic-main-outputs.cpp:
```diff
@@ -1381,6 +1384,9 @@ bool AdvancedOutput::StartStreaming(obs_service_t *service)
+        if (!codec) {
+            return false;
+        }
 
```
All fixed!

QuinnDamerell-MS commented on plugins/obs-ffmpeg/obs-ffmpeg-opus.c:
```diff
@@ -0,0 +1,303 @@
+/******************************************************************************
+    Copyright (C) 2014 by Quinn Damerell <qdamere@microsoft.com>
```
Fixed.

QuinnDamerell-MS commented on plugins/obs-ffmpeg/obs-ffmpeg-opus.c:
```diff
@@ -0,0 +1,303 @@
+{
+	int ret;
+
+	enc->aframe  = av_frame_alloc();
```
Fixed.

QuinnDamerell-MS commented on plugins/obs-ffmpeg/obs-ffmpeg-opus.c:
```diff
@@ -0,0 +1,303 @@
+{
+	struct opus_encoder *enc;
+	int                bitrate = (int)obs_data_get_int(settings, "bitrate");
+	audio_t            *audio   = obs_encoder_audio(encoder);
```
Fixed.

QuinnDamerell-MS commented on plugins/obs-ffmpeg/obs-ffmpeg-opus.c:
```diff
@@ -0,0 +1,303 @@
+	.get_properties = opus_properties,
+	.get_extra_data = opus_extra_data,
+	.get_audio_info = opus_audio_info
+};
```
Fixed.

QuinnDamerell-MS commented on plugins/obs-outputs/ftl-stream.c:
```diff
@@ -0,0 +1,1188 @@
+﻿/******************************************************************************
+    Copyright (C) 2014 by Quinn Damerell <qdamere@microsoft.com>
```
Fixed.

QuinnDamerell-MS commented on plugins/obs-outputs/ftl-stream.c:
```diff
@@ -0,0 +1,1188 @@
+    if (is_header) {
+        bfree(packet->data);
+    }
+    else {
```
Fixed.

QuinnDamerell-MS commented on plugins/obs-outputs/ftl-stream.c:
```diff
@@ -0,0 +1,1188 @@
+    // even if the speed test reported it should be lower. If we don't, FTL
+    // will queue data on the client and start adding latency. If the internet
+    // connection really can't handle the bitrate the user will see either lost frame
+    // and recovered frame counts go up, which is reflect in the dropped_frames count.
```
Fixed.

QuinnDamerell-MS commented on plugins/obs-outputs/ftl-stream.c:
```diff
@@ -0,0 +1,1188 @@
+
+		if (stopping(stream) && stream->stop_ts == 0) {
+			break;
+		}
```
Fixed.

QuinnDamerell-MS commented on plugins/obs-outputs/ftl-stream.c:
```diff
@@ -0,0 +1,1188 @@
+		}
+
+		/* sends sps/pps on every key frame as this is typically
+		 * required for webrtc */
```
Fixed.

QuinnDamerell-MS commented on plugins/obs-outputs/ftl-stream.c:
```diff
@@ -0,0 +1,1188 @@
+		info("Disconnected from %s", stream->path.array);
+	} else {
+		info("User stopped the stream");
+	}
```
Fixed.

QuinnDamerell-MS commented on plugins/obs-outputs/ftl-stream.c:
```diff
@@ -0,0 +1,1188 @@
+	status_code = ftl_ingest_disconnect(&stream->ftl_handle);
+	if (status_code != FTL_SUCCESS) {
+		printf("Failed to disconnect from ingest %d", status_code);
+	}
```
Fixed.

QuinnDamerell-MS commented on plugins/obs-outputs/ftl-stream.c:
```diff
@@ -0,0 +1,1188 @@
+	uint8_t       *header;
+	size_t        size;
+
+	struct encoder_packet packet   = {
```
Fixed.

QuinnDamerell-MS commented on plugins/obs-outputs/ftl-stream.c:
```diff
@@ -0,0 +1,1188 @@
+		.type         = OBS_ENCODER_VIDEO,
+		.timebase_den = 1,
+		.keyframe     = true,
+		.dts_usec = dts_usec
```
Fixed.

QuinnDamerell-MS commented on plugins/obs-outputs/ftl-stream.c:
```diff
@@ -0,0 +1,1188 @@
+
+    if (!obs_output_can_begin_data_capture(stream->output, 0)) {
+        return false;
+    }
```
Fixed.

QuinnDamerell-MS commented on plugins/obs-outputs/ftl-stream.c:
```diff
@@ -0,0 +1,1188 @@
+		added_packet = (packet->type == OBS_ENCODER_VIDEO) ?
+			add_video_packet(stream, &new_packet) :
+			add_packet(stream, &new_packet);
+	}
```
Fixed.

QuinnDamerell-MS commented on plugins/obs-outputs/ftl-stream.c:
```diff
@@ -0,0 +1,1188 @@
+
+	if (status.msg.event.reason == FTL_STATUS_EVENT_REASON_API_REQUEST) {
+		return RET_BREAK;
+	}
```
Fixed.

QuinnDamerell-MS commented on plugins/obs-outputs/ftl-stream.c:
```diff
@@ -0,0 +1,1188 @@
+			continue;
+		} else if (status_code == FTL_NOT_INITIALIZED) {
+			break;
+		}
```
Fixed.

QuinnDamerell-MS commented:
@Xaymar @notr1ch @derrod @RytoEX 

Great feedback! Thanks so much! I think I have addressed all of the issues, if you see any more please let me know!

RytoEX commented:
@QuinnDamerell-MS 
Did you happen to Find-and-Replace "&nbsp;&nbsp;&nbsp;&nbsp;" (4 spaces) with "&emsp;" (1 tab) across your entire patch, or most of it excluding the rtmp-services JSON?  That's what it seems like you did, and now you've introduced whitespace changes on line not originally included in this PR where there were no changes requested.  In fact, you've replaced spaces where Jim seems to have intentionally placed them, and you now have some lines tab-indented further than they should be indented.

I suggest that you revert your recent code style commit, and individually reapply the code style fixes that were previously requested.

You also haven't addressed the comments about the rtmp-services JSON [here](https://github.com/jp9000/obs-studio/pull/980#discussion_r130211707) or my questions about OBS-FTL's UI components for Opus [here](https://github.com/jp9000/obs-studio/pull/980#pullrequestreview-53067252).

jp9000 commented:
Seems he did do a find/replace.  Yea, that fixup commit will have to be removed.  It's not a huge deal though, I can fix those things myself during merge.  For the format version change, @derrod was correct in explaining what it was, the Json format for that specific sub-object for mixer had changed (no longer using RTMP urls), prompting a format version change because it wasn't backward compatible (because older versions of the program wouldx expect those RTMP urls).  However, because they also decided to change the name of the service, using the the format version increment is no longer necessary because it's now considered a separate new object, so I'll just remove that myself as well.

As for the audio codec stuff, the stuff that allows alternate audio codecs for streams was already in the UI prior to the PR.  The part at the bottom where it specifies encoded_audio_codecs = "opus" will cause it to use the Opus encoder automatically instead of AAC (for just the stream encoder).  I had talked with them about that before, that was one of the primary problems they originally had apparently.

I'm going to run it through some testing tomorrow and then merge along with a few other pending PRs.

RytoEX commented:
@jp9000
Thanks for the insight!  As I'd mentioned earlier, I'd looked into the differences between this and the OBS-FTL fork.  I had largely been referencing [the GitHub diff between the forks](https://github.com/jp9000/obs-studio/compare/master...mixer:ftl), so I had been focusing on the differences in `UI/audio-encoders.cpp` and `UI/audio-encoders.hpp` when I mentioned the UI bits of Opus support.  Glad to hear you're way ahead of me here!  As we discussed earlier, I'll stop reviewing this PR since you're going to handle the edits.

pkviet commented:
hi, the cutoff frequency in obs-ffmpeg-opus.c (line 162, which was copy pasted from ffmpeg-aac ) is quite probably off the mark for libopus; from libavcodec/libopusenc (function libopus_encode_init):

> Cutoff frequency must be exactly one of: 4000, 6000, 8000, 12000 or 20000

So it will probably be inactive most of the time.

jp9000 commented:
Okay, I've refactored obs-ffmpeg-aac to allow Opus as well, because the Opus encoder was basically a duplicate of the AAC encoder.  The FTL output is basically from the RTMP output, but with more significant changes, so I'm going to keep that separate rather than try to refactor the RTMP output.  I fixed the whitespace issues and cleaned up the commit messages.

A good tip for the future to check to see if your commits look good is to use "git diff" on the changes before committing, it'll usually point out any inconsistencies in the code formatting.

Fixed it up and will be merging with those changes shortly.  Will show up as "closed from [x commit]".
